File: lib/specialsExport.ts
// Pure CSV builder and helpers for the specials export pipeline.
// No I/O, no DB. The route layer is responsible for read/write side-effects.

export interface IngredientRow {
  ingredient: string;
  qty: number | string;
  unit: string;
  vendor_match: string;
  note: string;
}

export interface RecipeRow {
  slug: string;
  display_name: string;
  yield_qty: number;
  yield_unit: string;
  category: string;
  procedure: string;
}

const UNMATCHED_NOTE = 'unmatched — pick a vendor item before paste';

const RECIPE_HEADER = 'slug,display_name,yield_qty,yield_unit,category,procedure';
const INGREDIENT_HEADER = 'ingredient,qty,unit,vendor_match,note';

export function escapeCsvField(value: unknown): string {
  if (value === null || value === undefined) return '';
  const s = typeof value === 'string' ? value : String(value);
  if (/[",\n\r]/.test(s)) {
    return `"${s.replace(/"/g, '""')}"`;
  }
  return s;
}

function joinRow(fields: unknown[]): string {
  return fields.map(escapeCsvField).join(',');
}

export function mapCostBreakdownToIngredientRows(breakdown: unknown): IngredientRow[] {
  if (!Array.isArray(breakdown)) return [];
  return breakdown.map((row: any) => {
    const matched = typeof row?.match === 'string' && row.match.length > 0 && row?.cost !== null && row?.cost !== undefined;
    return {
      ingredient: typeof row?.item === 'string' ? row.item : '',
      qty: row?.req_qty ?? '',
      unit: typeof row?.req_unit === 'string' ? row.req_unit : '',
      vendor_match: matched ? row.match : '',
      note: matched ? '' : UNMATCHED_NOTE,
    };
  });
}

export function selectSkippedRows(rows: IngredientRow[]): IngredientRow[] {
  return rows.filter((r) => r.note === UNMATCHED_NOTE);
}

// Strip a trailing GitHub-style markdown blockquote (> [!NOTE] / > [!WARNING])
// emitted by the cost_special action handler. Anything before that block is
// kept verbatim — chefs may want it as procedure prose.
export function stripCostMarkdown(answer: string): string {
  if (typeof answer !== 'string') return '';
  const idx = answer.search(/\n\n> \[!(NOTE|WARNING)\]/);
  if (idx < 0) return answer;
  return answer.slice(0, idx).trimEnd();
}

export function buildExportCsv(input: { recipe_row: RecipeRow; ingredient_rows: IngredientRow[] }): string {
  const r = input.recipe_row;
  const recipeBody = joinRow([r.slug, r.display_name, r.yield_qty, r.yield_unit, r.category, r.procedure]);
  const ingredientBody = input.ingredient_rows
    .map((row) => joinRow([row.ingredient, row.qty, row.unit, row.vendor_match, row.note]))
    .join('\n');
  const tail = ingredientBody.length > 0 ? `${ingredientBody}\n` : '';
  return `# RECIPE\n${RECIPE_HEADER}\n${recipeBody}\n\n# INGREDIENTS\n${INGREDIENT_HEADER}\n${tail}`;
}

File: LariatNative/Sources/LariatModel/Compute/SpecialsExport.swift
import Foundation

/// Pure CSV builder for the specials export pipeline — parity with
/// `lib/specialsExport.ts`. No I/O, no DB.
public enum SpecialsExport {
    public static let unmatchedNote = "unmatched — pick a vendor item before paste"

    static let recipeHeader = "slug,display_name,yield_qty,yield_unit,category,procedure"
    static let ingredientHeader = "ingredient,qty,unit,vendor_match,note"

    public struct IngredientRow: Sendable, Equatable {
        public let ingredient: String
        /// Pre-stringified (JS `String(number)` semantics) — "" when absent.
        public let qty: String
        public let unit: String
        public let vendorMatch: String
        public let note: String

        public init(ingredient: String, qty: String, unit: String, vendorMatch: String, note: String) {
            self.ingredient = ingredient
            self.qty = qty
            self.unit = unit
            self.vendorMatch = vendorMatch
            self.note = note
        }
    }

    public struct RecipeRow: Sendable, Equatable {
        public let slug: String
        public let displayName: String
        public let yieldQty: Double
        public let yieldUnit: String
        public let category: String
        public let procedure: String

        public init(slug: String, displayName: String, yieldQty: Double,
                    yieldUnit: String, category: String, procedure: String) {
            self.slug = slug
            self.displayName = displayName
            self.yieldQty = yieldQty
            self.yieldUnit = yieldUnit
            self.category = category
            self.procedure = procedure
        }
    }

    // MARK: escapeCsvField

    /// RFC-4180: quote when the field contains `"`, `,`, `\n`, or `\r`;
    /// double embedded quotes.
    public static func escapeCsvField(_ value: String?) -> String {
        guard let s = value else { return "" }
        if s.contains("\"") || s.contains(",") || s.contains("\n") || s.contains("\r") {
            return "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\""
        }
        return s
    }

    static func joinRow(_ fields: [String?]) -> String {
        fields.map(escapeCsvField).joined(separator: ",")
    }

    // MARK: mapCostBreakdownToIngredientRows

    /// A line is "matched" iff `match` is a non-empty string AND `cost` is
    /// non-null (web truthiness check).
    public static func mapCostBreakdownToIngredientRows(_ breakdown: [CostBreakdownLine]) -> [IngredientRow] {
        breakdown.map { line in
            let matched = (line.match?.isEmpty == false) && line.cost != nil
            return IngredientRow(
                ingredient: line.item ?? "",
                qty: line.reqQtyString ?? "",
                unit: line.reqUnit ?? "",
                vendorMatch: matched ? (line.match ?? "") : "",
                note: matched ? "" : unmatchedNote
            )
        }
    }

    public static func selectSkippedRows(_ rows: [IngredientRow]) -> [IngredientRow] {
        rows.filter { $0.note == unmatchedNote }
    }

    // MARK: stripCostMarkdown

    /// Strip a trailing GitHub-style `> [!NOTE]` / `> [!WARNING]` blockquote
    /// emitted by the cost action handler; keep everything before it,
    /// trailing-whitespace-trimmed.
    public static func stripCostMarkdown(_ answer: String) -> String {
        guard let range = answer.range(
            of: "\\n\\n> \\[!(NOTE|WARNING)\\]",
            options: .regularExpression
        ) else { return answer }
        var head = String(answer[..<range.lowerBound])
        while let last = head.unicodeScalars.last,
              CharacterSet.whitespacesAndNewlines.contains(last) {
            head.unicodeScalars.removeLast()
        }
        return head
    }

    // MARK: buildExportCsv

    public static func buildExportCsv(recipeRow r: RecipeRow, ingredientRows: [IngredientRow]) -> String {
        let recipeBody = joinRow([
            r.slug, r.displayName, JsValueFormat.numberString(r.yieldQty),
            r.yieldUnit, r.category, r.procedure,
        ])
        let ingredientBody = ingredientRows
            .map { joinRow([$0.ingredient, $0.qty, $0.unit, $0.vendorMatch, $0.note]) }
            .joined(separator: "\n")
        let tail = ingredientBody.isEmpty ? "" : "\(ingredientBody)\n"
        return "# RECIPE\n\(recipeHeader)\n\(recipeBody)\n\n# INGREDIENTS\n\(ingredientHeader)\n\(tail)"
    }
}
